rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+        external_json_writer_ = std::make_shared<ExternalJsonWriterCB>();
+        ICHECK(external_json_writer_);
+      }
+      static ExternalJsonWriterCB* GetExternalJsonWriter() { return external_json_writer_.get(); }
+      ....
+      LoweredOutput Codegen(IRModule mod, relay::Function func, String mod_name) {
+        ....
+
+        // if it has been registered for this GraphExecutorCodegen object, call the external JSON writer
+        if (external_json_writer_->HasCallback()) {
+          std::ostringstream external_os;
+          dmlc::JSONWriter external_writer(&external_os);
+          external_json_writer_->Exe(&external_writer, ret.external_mods, nodes_, heads_);
+          ret.external_graph_json = external_os.str();
+        }
+
+        return ret;
+      }
+    };
+
+    extern "C" ExternalJsonWriterCB* GetExternalJsonWriter() {
+      return GraphExecutorCodegen::GetExternalJsonWriter();
+    }
+
+    /////////////
+    // in the new src/relay/backend/contrib/mrvl/graph_executor_codegen.cc file
+    // Marvell-specific extentions
+    class GraphInputNodeMrvlExt : public GraphInputNode {
+        ...
+        GraphNodeType Type() const override { return kGraphInputNodeExt; }
+        void Save(dmlc::JSONWriter* writer) const override { /* extensions */ }
+    }
+
+    class GraphOpNodeMrvlExt : public GraphOpNode {
+        ...
+        GraphNodeType Type() const override { return kGraphOpNodeExt; }
+        void Load(dmlc::JSONReader* reader) override;
+        void LoadAttrs(dmlc::JSONReader* reader);
+        std::pair<std::string, GraphAttrs> GetLoadedGraphAttrs();
+    }
+
+    class MrvlExtJson {
+     public:
+      MrvlExtJson() {
+        ICHECK(!GetExternalJsonWriter()->HasCallback()) << "ERROR: has registered callback";
+        GetExternalJsonWriter()->RegisterCB(this, &MrvlExtJson::GetExternalJSON);
+      }
+      virtual ~MrvlExtJson() {}
+      void GetExternalJSON(dmlc::JSONWriter* writer, Array<tvm::runtime::Module> external_mods,
+                           std::vector<GraphObjectPtr> nodes, std::vector<GraphNodeRef> heads);
+      void LoadExternalJsonAttrs(std::unordered_map<std::string, GraphAttrs>* external_attrs_map,
+                                 const Array<tvm::runtime::Module>& external_mods);
+    };
+```
+
+* the need to link between pre-trained model and final Marvell backend layer - for instance, through tvm_custom
+    * We did not include prototype code in PR-9730 but intend to provide our sample changes in another RFC and PR.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* We haven't identified any major *not* do items. Several other designs are by choices - that is we understand that
+  there are benefits for doing or benefits for not-doing.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* We follow the TVM BYOC framework to enable BYOC Marvell flow without impacting any TVM core features.
```

Comment:
it seems like there has been some impact to the GraphExecutor, and I think one point of confusion here is whether it was necessary to do that or whether you could have handled the additional runtime complexity inside a Marvell-specific `runtime.Module`. could you explain a bit further here?

Reply:
the email version of this question was linking to a different RFC segment (the light green section above) -- led to me to answer differently in my in-line reply to your email.
Sorry and since I noticed here and can see the correct light green block above corresponding to your question, so let me reply here again properly.

Please check my reply to the previous question (e.g., we need Marvell-specific GraphOpNode and GraphInputNode in order to dump out Marvell specific attributes to node-JSON file).
Also, since we are also using our build-from-outside-typical TVM flow Marvell compiler backend component, which can do additional compile-time optimizations and is reading in graph-executor-generated JSON meta data, currently, we don't think using runtime.Module to generate Marvell-specific JSON meta files of network is the way to go here.

Reply:
@areusch: We may not know enough but like to learn more about the runtime.Module subject (its flow and how to add specialization). Can you provide an example or an existing tvm/tests-folder suite as a pointer for us to run an existing runtime using tvm repo (without relying on specific USE_<vendor> flags being ON), if there is such example?